### PR TITLE
fix(api): strip header-only basepath for proxied callback routes (#2778)

### LINKED ISSUE #2778: bug: Overview not loading after latest nightly update, login doesnt work
### BirdNET-Go version / build date

nightly-20260414

### Install method

Unraid Community App

### Device type

Custom Server

### Operating system and version

OS: Unraid 7.2.4

### Hardware details

Enough

### Description

The Overview doesnt work after latest upgrade. Infererence still working. Also login isnt working after the update

Image

### Steps to reproduce

hard to tell, I just upgraded from the nightly before and had this issue

### Expected behaviour

Overview showing what was heard over the day, day selection working

### Actual behaviour

As shown in description, doesnt work directly in LAN and also not over reverse proxy

### Relevant log output

```shell
Container Log


INFO  [main] BirdNET-Go starting version=nightly-20260414 build_date=2026-04-14T08:38:02Z config_file=/home/birdnet/.config/birdnet-go/config.yaml
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
INFO  [analysis.processor] False positive filtering is disabled current_level=0 operation=false_positive_filter_config
INFO  [analysis.processor] Species tracking enabled window_days=7 sync_interval_minutes=60 hemisphere=northern latitude=49.43 operation=species_tracking_config
INFO  [analysis.processor] Loading dynamic thresholds from database operation=load_dynamic_thresholds
INFO  [analysis.processor] Dynamic thresholds loaded from database loaded_count=16 expired_count=0 total_retrieved=16 operation=load_dynamic_thresholds
INFO  [analysis.processor] Starting dynamic threshold persistence persist_interval_seconds=30 operation=threshold_persistence_startup
INFO  [analysis.processor] Starting dynamic threshold cleanup cleanup_interval_hours=24 operation=threshold_cleanup_startup
INFO  [analysis.processor] Daylight filter enabled for filtered species species_count=254 offset_hours=0 operation=daylight_filter_init
INFO  [notification-integration] initializing notification event bus integration
INFO  [notification-integration] notification worker registered with event bus consumer=notification-worker supports_batching=false batching_enabled=false batch_size=10 batch_timeout=100ms circuit_breaker_threshold=5 recovery_timeout=30s debug=false
INFO  [notification-integration] detection notification consumer registered with event bus consumer=detection-notification-consumer debug=false
INFO  [analysis.processor] Processor.Start() called — BufferMgr and Registry wired, launching detection goroutines buffer_mgr_set=true registry_set=true operation=processor_start
INFO  [analysis.processor] Starting detection processor operation=detection_processor_startup
INFO  [analysis.processor] Starting worker pool max_capacity=1000 component=analysis.processor
INFO  [analysis.processor] Worker pool started successfully max_capacity=1000 component=analysis.processor operation=worker_pool_start
INFO  [analysis.processor] Starting pending detections flusher flush_interval_seconds=1 operation=pending_flusher_startup
INFO  [analysis.processor] Created new pending detection species=türkentaube confidence=0.49 source=rtsp_764c4fad model_id=BirdNET_V2.4 flush_deadline=2026-04-15T13:10:43+02:00 operation=create_pending_detection
INFO  [analysis.processor] Flushing detection species=türkentaube source=M5Atom model_id=BirdNET_V2.4 deadline_reached=true count=7 required=1 operation=flush_detection
INFO  [analysis.processor] approving detection species=türkentaube source=M5Atom model_id=BirdNET_V2.4 match_count=7 confidence=0.91 operation=approve_detection
INFO  [analysis.processor] Audio clip saved successfully component=analysis.processor.actions detection_id=TÜR_1310_124c clip_path=2026/04/streptopelia_decaocto_91p_20260415T131043Z.flac file_size_bytes=105647 format=flac operation=audio_export_success
INFO  [analysis.processor] Created new pending detection species=türkentaube confidence=0.99 source=rtsp_764c4fad model_id=BirdNET_V2.4 flush_deadline=2026-04-15T13:10:57+02:00 operation=create_pending_detection
INFO  [analysis.processor] Flushing detection species=türkentaube source=M5Atom model_id=BirdNET_V2.4 deadline_reached=true count=3 required=1 operation=flush_detection
INFO  [analysis.processor] approving detection species=türkentaube source=M5Atom model_id=BirdNET_V2.4 match_count=3 confidence=0.99 operation=approve_detection
INFO  [analysis.processor] Audio clip saved successfully component=analysis.processor.actions detection_id=TÜR_1310_5766 clip_path=2026/04/streptopelia_decaocto_99p_20260415T131045Z.flac file_size_bytes=427594 format=flac operation=audio_export_success
```

### RTSP source (only if the bug is RTSP related)

_No response_

### Pre-submission checks

- [x] I am running the latest release or have tested with the `:dev` container image.
- [x] I have searched existing open and closed issues and this does not appear to be a duplicate.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tphakala**
## Summary
- Resolve the effective basepath inside the Echo Pre strip middleware via the shared `ingressPath()` helper, so requests that carry a basepath only in `X-Ingress-Path` or `X-Forwarded-Prefix` (no YAML `webserver.basepath`) get their prefix stripped and land on handlers.
- Fixes #2778: after PR #2731, the login handler emits callback URLs like `/birdnet/api/v2/auth/callback?...` derived from the same header-priority chain, but without this fix the returned URL 404s because the strip middleware only consulted `settings.WebServer.BasePath`. Reverse-proxy users who had not also set `webserver.basepath` saw the dashboard hang (login never completed) and the Overview never loaded.
- Preserves the existing short-circuit that skips stripping when a proxy header is present and the path does not start with the basepath (so proxies that already strip the prefix keep working), and keeps the `isSafePathPrefix` header guard from #2731 untouched.

## Validation
- QA run against a locally built fix image probed the bug scenarios and a baseline image. `X-Forwarded-Prefix: /birdnet` on `/birdnet/api/v2/health` flipped from 404 on baseline to 200 on the fix. Same flip for `X-Ingress-Path`. Login round-trip (POST `/api/v2/auth/login` with the proxy header, follow the returned `redirectUrl`) flipped from 404 to 302 to `/birdnet/ui/dashboard`.
- The existing 26-test Playwright basepath suite (`e2e/tests/basepath.spec.js`) still passes for the YAML-basepath scenario. One unrelated flake surfaced in the dashboard hourly-grid anchors (misses the basepath when detections exist) and is tracked separately; it is a frontend component bug, not a regression from this PR.

## Test plan
- [x] `go test -race ./internal/api/...` (includes new cases in `internal/api/basepath_test.go` for header-only basepath with both `X-Forwarded-Prefix` and `X-Ingress-Path`, the proxy-already-stripped short-circuit, and the fully empty case).
- [x] `golangci-lint run -v` on the full project - 0 issues.
- [x] `coderabbit review --plain --base main` - no findings.
- [x] Reverse-proxy simulation validated in QA (probes + login round-trip).
- [ ] Manual smoke: behind a real reverse proxy (Caddy/nginx/traefik) confirm dashboard loads and login completes.
- [ ] Manual regression check: `webserver.basepath` set in YAML still works without proxy headers.

Fixes #2778

## Summary by CodeRabbit

* **Tests**
  * Expanded test coverage for base-path handling with additional scenarios covering proxy headers, config edge cases, and skip conditions.

* **Bug Fixes**
  * Preserve percent-encoded paths when base-path stripping occurs.
  * Ensure proxy headers are honored in the correct precedence when determining the base path.

* **Chores**
  * Compute base-path per-request for more accurate handling of proxied requests.

### CODE REVIEW

File: internal/api/server.go [hunk 1]
```diff
@@ -283,25 +283,28 @@ func (s *Server) setupMiddleware() {
...
 	//
-	// The effective basepath is read per-request from s.settings so changes to
-	// settings.WebServer.BasePath take effect without a server restart, matching
-	// the dynamic behavior of ingressPath() used by the context middleware below.
+	// Resolving per-request also means settings.WebServer.BasePath changes take
+	// effect without a server restart.
 	s.echo.Pre(func(next echo.HandlerFunc) echo.HandlerFunc {
 		return func(c echo.Context) error {
-			bp := strings.TrimRight(s.settings.WebServer.BasePath, "/")
+			bp := ingressPath(c, s.settings)
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/api/server.go:298]: Pushback on this one. Verified in the codebase before changing:

1. `settingsInstance` in `internal/conf/config.go` is assigned exactly once at startup (`Load()` -> `settingsInstance = settings`). Runtime UI edits mutate fields on that live `*Settings` in place; the pointer itself is never swapped. Calls through `SaveSettings()` snapshot-copy the instance only to write YAML and never reassign the singleton. Test-only `SetTestSettings` is the one exception.
2. Therefore `s.settings` captured in the `Server` struct at startup continues to see current values, including for `WebServer.BasePath`. The context middleware directly below this one already relies on exactly this pattern (`ingressPath(c, s.settings)`), as does `auth/middleware.go:requestBasePath`.
3. The suggested `c.Settings` does not exist on the project's Echo context. Per-request settings access in this codebase goes through `c.Get(\"basePath\")` (populated by the context middleware above), not through a `c.Settings` field.

Net: no hot-reload regression introduced; the PR keeps the exact settings-access pattern the surrounding code uses. Leaving as-is.

**Comment by tphakala** [comment]
@coderabbitai review

**Comment by tphakala** [comment]
Addressed CodeRabbit's outside-diff finding in e1bebc373253b377d6f06ef902b4601728074cc4: the test helper at `internal/api/basepath_test.go` now mirrors production's `req.URL.RawPath` rewrite alongside `req.URL.Path`, so percent-encoded path regressions cannot slip past the test helper.

File: internal/api/basepath_test.go [hunk 2]
```diff
@@ -118,20 +142,56 @@ func TestBasePathStripMiddleware(t *testing.T) {
...
+			expectedPath:     "/api/v2/health",
+		},
+		{
+			// No source of basepath at all: neither header nor config is set,
+			// so the middleware must be a no-op.
+			name:         "no strip when no basepath from any source",
+			basePath:     "",
+			requestPath:  "/api/v2/health",
+			expectedPath: "/api/v2/health",
+		},
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/api/basepath_test.go:180]: Pushback. The whole point of the previous fix round (commit e1bebc37, addressing CodeRabbit's note) was to align the test helper with production behaviour. Production middleware in `internal/api/server.go` does the RawPath rewrite via manual string manipulation, so the test helper has to do the same to keep test/production parity; if the test helper used `net/url` while production used string slicing, any percent-encoding regression in production would slip past the helper.

Refactoring both production and test helper to `net/url` would be a sensible cleanup but it is a separate PR and out of scope for this hotfix (which is targeted at the header-only basepath strip bug in #2778). Leaving the helper as-is for parity.
